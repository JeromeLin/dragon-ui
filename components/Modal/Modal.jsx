import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import Events from '../utils/events';

class Modal extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isShow: false,
      isPending: false,
      animationState: 'leave'
    };

    this.animationEnd = this.animationEnd.bind(this);
  }

  componentWillMount() {
    if (this.props.visible) {
      this.enter();
    }
  }

  componentWillUpdate() {
    Events.on(this.modal, 'webkitAnimationEnd', this.animationEnd);
    Events.on(this.modal, 'animationend', this.animationEnd);
  }

  componentWillUnmount() {
    Events.off(this.modal, 'webkitAnimationEnd', this.animationEnd);
    Events.off(this.modal, 'animationend', this.animationEnd);
  }

  componentWillReceiveProps(nextProps) {
    if (!this.props.visible && nextProps.visible) {
      this.enter();
    } else if (this.props.visible && !nextProps.visible) {
      this.leave();
    }
  }

  shouldComponentUpdate(nextProps, nextState) {
    return !!(this.state.isShow || nextState.isShow);
  }

  render() {
    const {
      animationType,
      animationDuration,
      width,
      minWidth,
      isRadius,
      isRound,
      className,
      onMaskClick,
      children,
    } = this.props;
    const { isShow, isPending, animationState } = this.state;

    const classes = {
      modal: classnames({
        'ui-modal': true,
        radius: 'radius' in this.props || isRadius,
        round: 'round' in this.props || isRound,
        [`fade-${animationState}`]: isPending,
        [className]: !!className
      }),
      dialog: classnames({
        'ui-modal-dialog': true,
        [`${animationType}-${animationState}`]: true
      })
    };

    const style = {
      modal: {
        WebkitAnimationDuration: `${animationDuration}ms`,
        MozAnimationDuration: `${animationDuration}ms`,
        msAnimationDuration: `${animationDuration}ms`,
        OAnimationDuration: `${animationDuration}ms`,
        animationDuration: `${animationDuration}ms`
      },
      dialog: {
        width,
        minWidth,
        WebkitAnimationDuration: `${animationDuration}ms`,
        MozAnimationDuration: `${animationDuration}ms`,
        msAnimationDuration: `${animationDuration}ms`,
        OAnimationDuration: `${animationDuration}ms`,
        animationDuration: `${animationDuration}ms`
      }
    };

    if (!isShow) {
      style.modal.display = 'none';
    }

    return (
      <div
        className={classes.modal}
        style={style.modal}
        onClick={onMaskClick}
        ref={(ele) => {
          this.modal = ele;
        }}
      >
        <div className="ui-modal-wrapper">
          <div
            className={classes.dialog}
            style={style.dialog}
            onClick={e => e.stopPropagation()}
          >
            {children}
          </div>
        </div>
      </div>
    );
  }

  animationEnd() {
    if (this.state.animationState === 'leave') {
      this.setState({
        isShow: false,
        isPending: false
      });
    } else {
      this.setState({
        isShow: true,
        isPending: false
      });
    }
  }

  enter() {
    this.setState({
      isShow: true,
      isPending: true,
      animationState: 'enter'
    });
  }

  leave() {
    this.setState({
      isShow: true,
      isPending: true,
      animationState: 'leave'
    });
  }
}

Modal.propTypes = {
  visible: PropTypes.bool,
  animationType: PropTypes.oneOf([
    'fade',
    'door',
    'flip',
    'rotate',
    'zoom',
    'moveUp',
    'moveDown',
    'moveLeft',
    'moveRight',
    'slideUp',
    'slideDown',
    'slideLeft',
    'slideRight'
  ]),
  animationDuration: PropTypes.number,
  width: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  minWidth: PropTypes.number,
  isRadius: PropTypes.bool,
  isRound: PropTypes.bool,
  onMaskClick: PropTypes.func
};

Modal.defaultProps = {
  visible: false,
  animationType: 'zoom',
  animationDuration: 300,
  width: 600,
  minWidth: 270,
  isRadius: false,
  isRound: false,
  onMaskClick() {}
};

export default Modal;
