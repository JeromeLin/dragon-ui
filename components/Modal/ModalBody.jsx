import React, { Component } from 'react';
import PropTypes from 'prop-types';

class ModalBody extends Component {
  render() {
    const { height, children } = this.props;
    const bodyStyle = {
      height
    };

    return (
      <div className="ui-modal-body" style={bodyStyle}>
        {children}
      </div>
    );
  }
}

ModalBody.propTypes = {
  height: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
};

ModalBody.defaultProps = {
  height: 'auto'
};

export default ModalBody;
