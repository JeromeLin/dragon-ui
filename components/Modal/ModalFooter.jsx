import React, { Component } from 'react';

class ModalFooter extends Component {
  render() {
    const { children, style } = this.props;

    return (
      <div className="ui-modal-footer" style={style}>
        {children}
      </div>
    );
  }
}

export default ModalFooter;
