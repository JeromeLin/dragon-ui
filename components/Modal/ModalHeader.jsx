import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Icon from '../Icon';

class ModalHeader extends Component {
  render() {
    const { title, onClose, style } = this.props;
    const btnClose = onClose ? (
      <div className="ui-modal-close" onClick={onClose}>
        <Icon type="wrong" />
      </div>
    ) : null;

    return (
      <div className="ui-modal-header" style={style}>
        <div className="ui-modal-title">{title}</div>
        {btnClose}
      </div>
    );
  }
}

ModalHeader.propTypes = {
  title: PropTypes.string,
  onClose: PropTypes.func,
  style: PropTypes.objectOf(PropTypes.any)
};

ModalHeader.defaultProps = {
  title: '',
  style: {},
  onClose: () => {},
};

export default ModalHeader;
