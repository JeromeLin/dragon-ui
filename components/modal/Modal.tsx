import React, { Component, KeyboardEventHandler } from 'react';
import Popup from 'zarm/lib/popup';
import cn from 'classnames';

import { ModalProps } from './PropsType';
import Button from '../button';

import ModalHeader from './ModalHeader';
import ModalBody from './ModalBody';
import ModalFooter from './ModalFooter';
import { Alert, Confirm } from './alert';

interface StateIF {
  isShow: boolean;
  sleep: boolean;
}

class Modal extends Component<ModalProps, StateIF> {
  static Alert = Alert;

  static Confirm = Confirm;

  static defaultProps = {
    prefixCls: 'zw-modal',
    okText: '确定',
    cancelText: '取消',
    closable: true,
    maskClosable: false,
    mask: true,
    centered: true,
    autoFocus: true,
    disableEscapeKeyDown: false,
    disableEnterKeyDown: false,
    zIndex: 2020,
    hideWhenShowOthers: true,
    destroy: false,
    scrollInModal: false,
    animationType: 'zoom',
    radius: true,
  };

  private static visibleList: Modal[] = [];

  private static activeElem: Element[] = [document.activeElement || document.body];

  private static handleVisbibleList(instance: Modal, visible: boolean) {
    if (visible) {
      const lastIndex = Modal.visibleList.length - 1;
      if (lastIndex >= 0) {
        const theLastInstance = Modal.visibleList[lastIndex];
        if (theLastInstance.props.hideWhenShowOthers === true) {
          theLastInstance.state.sleep = true;
          theLastInstance.setState({
            isShow: false,
          });
        }
      }
      Modal.visibleList.push(instance);
    } else {
      Modal.visibleList.pop();
      let index = Modal.visibleList.length;
      if (index > 0) {
        const modal = Modal.visibleList[index - 1];
        const currentVisible = modal.props.visible;
        if (currentVisible) {
          modal.setState({
            isShow: true,
          });
          modal.state.sleep = false;
        }
      }
      // eslint-disable-next-line no-plusplus
      while (index--) {
        const modal = Modal.visibleList[index];
        const currentVisible = modal.props.visible;
        if (!currentVisible) {
          modal.state.sleep = false;
          Modal.visibleList.splice(index, 1);
        }
      }
    }
  }

  static getDerivedStateFromProps(props: ModalProps, state: StateIF) {
    if (state.sleep) {
      return null;
    }
    return {
      isShow: !!props.visible,
    };
  }

  private modalContent!: HTMLDivElement;

  state: StateIF;

  constructor(props: ModalProps) {
    super(props);
    const { visible } = this.props;
    this.state = {
      isShow: !!visible,
      sleep: false,
    };
  }

  componentDidMount() {
    const { autoFocus, visible } = this.props;
    setTimeout(() => {
      if (this.modalContent && autoFocus) {
        if (visible) {
          Modal.handleVisbibleList(this, !!visible);
          if (this.modalContent && autoFocus) {
            Modal.activeElem.push(this.modalContent);
            this.modalContent.focus();
          }
        }
      }
    });
  }

  componentDidUpdate(prevProps: ModalProps) {
    const { autoFocus, visible } = this.props;
    if (prevProps.visible && visible === false) {
      Modal.handleVisbibleList(this, !!visible);
      setTimeout(() => {
        if (this.modalContent && autoFocus) {
          Modal.activeElem.pop();
          const lastElem = Modal.activeElem[Modal.activeElem.length - 1];
          if (lastElem instanceof HTMLElement) {
            lastElem.focus();
          }
        }
      });
    }
    if (visible && !prevProps.visible) {
      setTimeout(() => {
        if (this.modalContent && autoFocus) {
          Modal.activeElem.push(this.modalContent);
          this.modalContent.focus();
        }
      });
      Modal.handleVisbibleList(this, !!visible);
    }
  }

  setModalContainer = (elem: null | HTMLDivElement) => {
    if (elem) {
      this.modalContent = elem;
    }
  };

  onKeyDown: KeyboardEventHandler<HTMLDivElement> = (e) => {
    const { visible, onCancel, onKeyDown, disableEscapeKeyDown } = this.props;
    if (visible && !disableEscapeKeyDown) {
      if (e.keyCode === 27) {
        if (onCancel) {
          onCancel();
        }
      }
    }
    if (onKeyDown) {
      onKeyDown(e);
    }
  };

  onKeyPress: KeyboardEventHandler<HTMLDivElement> = (e) => {
    const { visible, onOk, onKeyPress, disableEnterKeyDown } = this.props;
    if (visible && disableEnterKeyDown === false) {
      if (this.modalContent === document.activeElement && visible && e.nativeEvent.keyCode === 13) {
        if (onOk) {
          onOk();
        }
      }
    }
    if (onKeyPress) {
      onKeyPress(e);
    }
  };

  onBlur = () => {
    this.modalContent.focus();
  };

  render() {
    const {
      prefixCls,
      children,
      title,
      closable,
      visible,
      zIndex,
      style,
      onOk,
      onCancel,
      okText,
      cancelText,
      className,
      centered,
      footer,
      scrollInModal,
      radius,
      ...others
    } = this.props;
    const styles = { ...style, zIndex };
    const { isShow } = this.state;
    const show = isShow;
    const hasFooter = footer !== null;
    const showHeader = title !== undefined || closable;
    const rewriteClassName = cn({
      [prefixCls]: true,
      [`${className}`]: !!className,
      [`${prefixCls}--popup`]: true,
      [`${prefixCls}--popup--top`]: !centered,
      [`${prefixCls}--scroll`]: scrollInModal,
      [`${prefixCls}--radius`]: radius,
    });
    return (
      <Popup
        visible={show}
        direction="center"
        {...others}
        className={rewriteClassName}
      >
        <div
          className={`${prefixCls}__content`}
          tabIndex={-1}
          onKeyDown={this.onKeyDown}
          onKeyPress={this.onKeyPress}
          ref={this.setModalContainer}
          style={styles}
          data-show={show}
          data-visible={visible}
        >
          {showHeader && <ModalHeader closable={closable} onCancel={onCancel}>{title}</ModalHeader>}
          <ModalBody>{children}</ModalBody>
          <ModalFooter>
            <div className={`${prefixCls}__button-warpper`}>
              {
                hasFooter && (footer || (
                  <>
                    <Button onClick={onCancel}>{cancelText}</Button>
                    <Button theme="primary" onClick={onOk}>{okText}</Button>
                  </>
                ))
              }
            </div>
          </ModalFooter>
        </div>
        <div className={`${prefixCls}__hidden-elem`}>
          <input type="readonly" tabIndex={0} onFocus={this.onBlur} />
        </div>
      </Popup>
    );
  }
}

export default Modal;
