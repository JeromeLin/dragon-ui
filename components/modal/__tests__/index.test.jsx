import React from 'react';
// import ReactDOM from 'react-dom';
import { render, shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Modal from '../index';

jest.mock('react-dom', () => ({
  createPortal: (node) => node,
}));

describe('Modal', () => {
  it('renders basic Modal correctly', () => {
    const wrapper = render(
      <Modal title="标题">
        我是一个模态框
      </Modal>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders customized Modal correctly', () => {
    const wrapper = render(
      <Modal
        width={400}
        radius
        visible
        animationType="fade"
        animationDuration={500}
      >
        我是一个模态框
      </Modal>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders closable Modal correctly', () => {
    const wrapper = render(
      <Modal
        title="标题"
        onCancel={() => { }}
      >
        我是一个模态框
      </Modal>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('triggers onClose callback correctly when clicks on closable icon', () => {
    const onClose = jest.fn();
    const wrapper = mount(
      <Modal
        visible
        title="标题"
        onCancel={onClose}
      >
        我是一个模态框
      </Modal>,
    );
    wrapper.find('.zw-modal__close').simulate('click');
    expect(onClose).toHaveBeenCalled();
  });

  it('triggers onMaskClick callback correctly', () => {
    const onMaskClick = jest.fn();
    const wrapper = shallow(
      <Modal
        title="标题"
        onMaskClick={onMaskClick}
      >
        我是一个模态框
      </Modal>,
    );

    wrapper.find('.za-mask').simulate('click');
    expect(onMaskClick).toHaveBeenCalled();
  });
});
