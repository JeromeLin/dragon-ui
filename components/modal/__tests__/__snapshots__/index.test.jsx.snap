// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal renders basic Modal correctly 1`] = `null`;

exports[`Modal renders closable Modal correctly 1`] = `null`;

exports[`Modal renders customized Modal correctly 1`] = `
Array [
  <div
    class="za-mask za-mask--normal"
    style="-webkit-animation-duration:500ms;animation-duration:500ms"
  />,
  <div
    class="za-popup__wrapper"
    role="dialog"
    style="-webkit-animation-duration:500ms;animation-duration:500ms"
  >
    <div
      class="za-popup za-popup--center"
      role="document"
      style="-webkit-animation-duration:500ms;animation-duration:500ms"
    >
      <div
        class="zw-modal__content"
        tabindex="-1"
      >
        <div
          class="zw-modal__header"
        >
          <div
            class="zw-modal__title"
          />
          <div
            class="zw-modal__close"
          >
            <i
              class="zw-icon zw-icon--wrong zw-icon--default zw-icon--sm"
              theme="default"
              type="wrong"
            >
              <svg
                fill="currentColor"
                height="1em"
                viewBox="0 0 32 32"
                width="1em"
              >
                <use
                  href="#wrong"
                />
              </svg>
            </i>
          </div>
        </div>
        <div
          class="zw-modal__body"
        >
          我是一个模态框
        </div>
        <div
          class="zw-modal__footer"
        >
          <div
            class="zw-modal__button-warpper"
          >
            <button
              class="zw-button zw-button--default zw-button--md zw-button--radius"
              type="button"
            >
              <span>
                取消
              </span>
            </button>
            <button
              class="zw-button zw-button--primary zw-button--md zw-button--radius"
              type="button"
            >
              <span>
                确定
              </span>
            </button>
          </div>
        </div>
      </div>
      <div
        class="zw-modal__hidden-elem"
      >
        <input
          tabindex="0"
          type="readonly"
        />
      </div>
    </div>
  </div>,
]
`;
